File: packages/agent-tools/src/policy.ts
import type { NormalConfig } from "@normal/schema";
import type { PatchOperation } from "./patch.js";

export type PolicyIssue = {
  readonly path: string;
  readonly code:
    | "path-not-writable"
    | "engine-managed-field"
    | "too-many-operations"
    | "weakens-attention-policy"
    | "sensitive-path";
  readonly severity: "deny" | "review";
  readonly message: string;
};

export type PolicyVerdict = {
  readonly denied: readonly PolicyIssue[];
  readonly review: readonly PolicyIssue[];
  readonly requiresApproval: boolean;
};

export const MAX_OPERATIONS = 64;

const WRITABLE_ROOTS = ["/spec", "/metadata/description", "/metadata/labels"];

const ENGINE_MANAGED = ["/apiVersion", "/kind", "/metadata/name", "/metadata/revision"];

const SENSITIVE_PATTERNS = [
  /^\/spec\/attention(\/|$)/,
  /^\/spec\/apps\/entries\/[^/]+\/(permissions|network|state)(\/|$)/,
  /^\/spec\/apps\/policy$/,
];

const covers = (root: string, path: string): boolean => path === root || path.startsWith(`${root}/`);

const isWritable = (path: string): boolean => WRITABLE_ROOTS.some((root) => covers(root, path));

const isEngineManaged = (path: string): boolean =>
  ENGINE_MANAGED.some((root) => covers(root, path));

const isSensitive = (path: string): boolean =>
  SENSITIVE_PATTERNS.some((pattern) => pattern.test(path));

const ENFORCEMENT_STRENGTH: Readonly<Record<string, number>> = {
  warn: 1,
  paginate: 2,
  block: 3,
};

const weakenings = (before: NormalConfig, after: NormalConfig): PolicyIssue[] => {
  const issues: PolicyIssue[] = [];
  const from = before.spec.attention.infiniteScroll;
  const to = after.spec.attention.infiniteScroll;

  const fromStrength = ENFORCEMENT_STRENGTH[from.enforcement] ?? 0;
  const toStrength = ENFORCEMENT_STRENGTH[to.enforcement] ?? 0;
  if (toStrength < fromStrength) {
    issues.push({
      path: "/spec/attention/infiniteScroll/enforcement",
      code: "weakens-attention-policy",
      severity: "review",
      message: `enforcement drops from '${from.enforcement}' to '${to.enforcement}'`,
    });
  }
  if (to.maxAutoLoads > from.maxAutoLoads) {
    issues.push({
      path: "/spec/attention/infiniteScroll/maxAutoLoads",
      code: "weakens-attention-policy",
      severity: "review",
      message: `automatic loads per session rise from ${from.maxAutoLoads} to ${to.maxAutoLoads}`,
    });
  }
  if (to.pageSize > from.pageSize) {
    issues.push({
      path: "/spec/attention/infiniteScroll/pageSize",
      code: "weakens-attention-policy",
      severity: "review",
      message: `page size grows from ${from.pageSize} to ${to.pageSize}`,
    });
  }
  const beforeExempt = new Set(from.exemptions.map((exemption) => exemption.package));
  const added = to.exemptions.filter((exemption) => !beforeExempt.has(exemption.package));
  for (const exemption of added) {
    issues.push({
      path: "/spec/attention/infiniteScroll/exemptions",
      code: "weakens-attention-policy",
      severity: "review",
      message: `adds a scroll exemption for '${exemption.package}' until ${exemption.expiresAt}`,
    });
  }
  if (from.webview.interceptIntersectionObserver && !to.webview.interceptIntersectionObserver) {
    issues.push({
      path: "/spec/attention/infiniteScroll/webview/interceptIntersectionObserver",
      code: "weakens-attention-policy",
      severity: "review",
      message: "stops intercepting the sentinel pattern web feeds use to auto-load",
    });
  }
  const beforeBudgets = new Map(
    before.spec.attention.sessionBudgets.map((budget) => [budget.id, budget]),
  );
  for (const budget of after.spec.attention.sessionBudgets) {
    const previous = beforeBudgets.get(budget.id);
    if (previous !== undefined && budget.dailyMinutes > previous.dailyMinutes) {
      issues.push({
        path: `/spec/attention/sessionBudgets/${budget.id}/dailyMinutes`,
        code: "weakens-attention-policy",
        severity: "review",
        message: `daily budget rises from ${previous.dailyMinutes} to ${budget.dailyMinutes} minutes`,
      });
    }
  }
  return issues;
};

export const checkOperationCount = (operations: readonly PatchOperation[]): PolicyIssue[] => {
  if (operations.length === 0) {
    return [
      {
        path: "/",
        code: "too-many-operations",
        severity: "deny",
        message: "a proposal must contain at least one operation",
      },
    ];
  }
  if (operations.length > MAX_OPERATIONS) {
    return [
      {
        path: "/",
        code: "too-many-operations",
        severity: "deny",
        message: `a proposal may contain at most ${MAX_OPERATIONS} operations`,
      },
    ];
  }
  return [];
};

export const checkOperationPaths = (operations: readonly PatchOperation[]): PolicyIssue[] => {
  const issues: PolicyIssue[] = [];
  for (const operation of operations) {
    if (isEngineManaged(operation.path)) {
      issues.push({
        path: operation.path,
        code: "engine-managed-field",
        severity: "deny",
        message: "this field is managed by the mutation engine and cannot be set by an agent",
      });
      continue;
    }
    if (!isWritable(operation.path)) {
      issues.push({
        path: operation.path,
        code: "path-not-writable",
        severity: "deny",
        message: `agents may only write under ${WRITABLE_ROOTS.join(", ")}`,
      });
      continue;
    }
    if (isSensitive(operation.path)) {
      issues.push({
        path: operation.path,
        code: "sensitive-path",
        severity: "review",
        message: "this path changes attention or app-permission policy and needs explicit approval",
      });
    }
  }
  return issues;
};

export const checkOperations = (operations: readonly PatchOperation[]): PolicyIssue[] => [
  ...checkOperationCount(operations),
  ...checkOperationPaths(operations),
];

export const checkPolicy = (
  operations: readonly PatchOperation[],
  before: NormalConfig,
  after: NormalConfig,
): PolicyVerdict => {
  const issues = [...checkOperationPaths(operations), ...weakenings(before, after)];
  const denied = issues.filter((issue) => issue.severity === "deny");
  const review = issues.filter((issue) => issue.severity === "review");
  return { denied, review, requiresApproval: review.length > 0 };
};

File: pkg/agent/policy.go
package agent

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/brandonapol/normal/pkg/config"
)

const MaxOperations = 64

type Severity string

const (
	SeverityDeny   Severity = "deny"
	SeverityReview Severity = "review"
)

type PolicyIssue struct {
	Path     string   `json:"path"`
	Code     string   `json:"code"`
	Severity Severity `json:"severity"`
	Message  string   `json:"message"`
}

type Verdict struct {
	Denied           []PolicyIssue
	Review           []PolicyIssue
	RequiresApproval bool
}

var writableRoots = []string{"/spec", "/metadata/description", "/metadata/labels"}

var engineManaged = []string{"/apiVersion", "/kind", "/metadata/name", "/metadata/revision"}

var sensitivePatterns = []*regexp.Regexp{
	regexp.MustCompile(`^/spec/attention(/|$)`),
	regexp.MustCompile(`^/spec/apps/entries/[^/]+/(permissions|network|state)(/|$)`),
	regexp.MustCompile(`^/spec/apps/policy$`),
}

func covers(root, path string) bool {
	return path == root || strings.HasPrefix(path, root+"/")
}

func isWritable(path string) bool {
	for _, root := range writableRoots {
		if covers(root, path) {
			return true
		}
	}
	return false
}

func isEngineManaged(path string) bool {
	for _, root := range engineManaged {
		if covers(root, path) {
			return true
		}
	}
	return false
}

func isSensitive(path string) bool {
	for _, pattern := range sensitivePatterns {
		if pattern.MatchString(path) {
			return true
		}
	}
	return false
}

var enforcementStrength = map[string]int{"warn": 1, "paginate": 2, "block": 3}

func weakenings(before, after config.Config) []PolicyIssue {
	issues := make([]PolicyIssue, 0)
	from := before.Spec.Attention.InfiniteScroll
	to := after.Spec.Attention.InfiniteScroll

	if enforcementStrength[to.Enforcement] < enforcementStrength[from.Enforcement] {
		issues = append(issues, PolicyIssue{
			Path:     "/spec/attention/infiniteScroll/enforcement",
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message:  fmt.Sprintf("enforcement drops from %q to %q", from.Enforcement, to.Enforcement),
		})
	}
	if to.MaxAutoLoads > from.MaxAutoLoads {
		issues = append(issues, PolicyIssue{
			Path:     "/spec/attention/infiniteScroll/maxAutoLoads",
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message: fmt.Sprintf("automatic loads per session rise from %d to %d",
				from.MaxAutoLoads, to.MaxAutoLoads),
		})
	}
	if to.PageSize > from.PageSize {
		issues = append(issues, PolicyIssue{
			Path:     "/spec/attention/infiniteScroll/pageSize",
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message:  fmt.Sprintf("page size grows from %d to %d", from.PageSize, to.PageSize),
		})
	}

	existing := make(map[string]bool, len(from.Exemptions))
	for _, exemption := range from.Exemptions {
		existing[exemption.Package] = true
	}
	for _, exemption := range to.Exemptions {
		if existing[exemption.Package] {
			continue
		}
		issues = append(issues, PolicyIssue{
			Path:     "/spec/attention/infiniteScroll/exemptions",
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message: fmt.Sprintf("adds a scroll exemption for %q until %s",
				exemption.Package, exemption.ExpiresAt),
		})
	}

	if from.Webview.InterceptIntersectionObserver && !to.Webview.InterceptIntersectionObserver {
		issues = append(issues, PolicyIssue{
			Path:     "/spec/attention/infiniteScroll/webview/interceptIntersectionObserver",
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message:  "stops intercepting the sentinel pattern web feeds use to auto-load",
		})
	}

	budgets := make(map[string]config.SessionBudget, len(before.Spec.Attention.SessionBudgets))
	for _, budget := range before.Spec.Attention.SessionBudgets {
		budgets[budget.ID] = budget
	}
	for _, budget := range after.Spec.Attention.SessionBudgets {
		previous, existed := budgets[budget.ID]
		if !existed || budget.DailyMinutes <= previous.DailyMinutes {
			continue
		}
		issues = append(issues, PolicyIssue{
			Path:     fmt.Sprintf("/spec/attention/sessionBudgets/%s/dailyMinutes", budget.ID),
			Code:     "weakens-attention-policy",
			Severity: SeverityReview,
			Message: fmt.Sprintf("daily budget rises from %d to %d minutes",
				previous.DailyMinutes, budget.DailyMinutes),
		})
	}

	return issues
}

func CheckOperationCount(operations []Operation) []PolicyIssue {
	if len(operations) == 0 {
		return []PolicyIssue{{
			Path: "/", Code: "invalid-operation-count", Severity: SeverityDeny,
			Message: "a proposal must contain at least one operation",
		}}
	}
	if len(operations) > MaxOperations {
		return []PolicyIssue{{
			Path: "/", Code: "invalid-operation-count", Severity: SeverityDeny,
			Message: fmt.Sprintf("a proposal may contain at most %d operations", MaxOperations),
		}}
	}
	return nil
}

func CheckOperationPaths(operations []Operation) []PolicyIssue {
	issues := make([]PolicyIssue, 0)
	for _, operation := range operations {
		switch {
		case isEngineManaged(operation.Path):
			issues = append(issues, PolicyIssue{
				Path: operation.Path, Code: "engine-managed-field", Severity: SeverityDeny,
				Message: "this field is managed by the mutation engine and cannot be set by an agent",
			})
		case !isWritable(operation.Path):
			issues = append(issues, PolicyIssue{
				Path: operation.Path, Code: "path-not-writable", Severity: SeverityDeny,
				Message: "agents may only write under " + joinMessages(writableRoots),
			})
		case isSensitive(operation.Path):
			issues = append(issues, PolicyIssue{
				Path: operation.Path, Code: "sensitive-path", Severity: SeverityReview,
				Message: "this path changes attention or app-permission policy and needs explicit approval",
			})
		}
	}
	return issues
}

func CheckOperations(operations []Operation) []PolicyIssue {
	return append(CheckOperationCount(operations), CheckOperationPaths(operations)...)
}

func CheckPolicy(operations []Operation, before, after config.Config) Verdict {
	issues := append(CheckOperationPaths(operations), SecurityRegressions(before, after)...)
	verdict := Verdict{Denied: make([]PolicyIssue, 0), Review: make([]PolicyIssue, 0)}
	for _, issue := range issues {
		if issue.Severity == SeverityDeny {
			verdict.Denied = append(verdict.Denied, issue)
			continue
		}
		verdict.Review = append(verdict.Review, issue)
	}
	verdict.RequiresApproval = len(verdict.Review) > 0
	return verdict
}
